core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala

```diff
@@ -0,0 +1,297 @@
+      "controller epoch was not persisted after broker failure")
+  }
+
+  @Test
+  def testTopicCreation(): Unit = {
+    servers = makeServers(1)
+    val tp = TopicAndPartition("t", 0)
+    val assignment = Map(tp.partition -> Seq(0))
+    TestUtils.createTopic(zkUtils, tp.topic, partitionReplicaAssignment = assignment, servers = servers)
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 0, LeaderAndIsr.initialLeaderEpoch,
+      "failed to get expected partition state upon topic creation")
+  }
+
+  @Test
+  def testTopicCreationWithOfflineReplica(): Unit = {
+    servers = makeServers(2)
+    servers(1).shutdown()
+    servers(1).awaitShutdown()
+    val tp = TopicAndPartition("t", 0)
+    val assignment = Map(tp.partition -> Seq(1, 0))
+    TestUtils.createTopic(zkUtils, tp.topic, partitionReplicaAssignment = assignment, servers = servers.take(1))
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 0, LeaderAndIsr.initialLeaderEpoch,
+      "failed to get expected partition state upon topic creation")
+  }
+
+  @Test
+  def testTopicPartitionExpansion(): Unit = {
+    servers = makeServers(1)
+    val tp0 = TopicAndPartition("t", 0)
+    val tp1 = TopicAndPartition("t", 1)
+    val assignment = Map(tp0.partition -> Seq(0))
+    val expandedAssignment = Map(tp0.partition -> Seq(0), tp1.partition -> Seq(0))
+    TestUtils.createTopic(zkUtils, tp0.topic, partitionReplicaAssignment = assignment, servers = servers)
+    zkUtils.updatePersistentPath(ZkUtils.getTopicPath(tp0.topic), zkUtils.replicaAssignmentZkData(expandedAssignment.map(kv => kv._1.toString -> kv._2)))
+    waitForPartitionState(tp1, KafkaController.InitialControllerEpoch, 0, LeaderAndIsr.initialLeaderEpoch,
+      "failed to get expected partition state upon topic partition expansion")
+    TestUtils.waitUntilMetadataIsPropagated(servers, tp1.topic, tp1.partition)
+  }
+
+  @Test
+  def testTopicPartitionExpansionWithOfflineReplica(): Unit = {
+    servers = makeServers(2)
+    val tp0 = TopicAndPartition("t", 0)
+    val tp1 = TopicAndPartition("t", 1)
+    val assignment = Map(tp0.partition -> Seq(1, 0))
+    val expandedAssignment = Map(tp0.partition -> Seq(1, 0), tp1.partition -> Seq(1, 0))
+    TestUtils.createTopic(zkUtils, tp0.topic, partitionReplicaAssignment = assignment, servers = servers)
+    servers(1).shutdown()
+    servers(1).awaitShutdown()
+    zkUtils.updatePersistentPath(ZkUtils.getTopicPath(tp0.topic), zkUtils.replicaAssignmentZkData(expandedAssignment.map(kv => kv._1.toString -> kv._2)))
+    waitForPartitionState(tp1, KafkaController.InitialControllerEpoch, 0, LeaderAndIsr.initialLeaderEpoch,
+      "failed to get expected partition state upon topic partition expansion")
+    TestUtils.waitUntilMetadataIsPropagated(servers.take(1), tp1.topic, tp1.partition)
+  }
+
+  @Test
+  def testPartitionReassignment(): Unit = {
```

Comment:
Same as above:
There is some small amount of overlap in terms of the protocols covered. ControllerIntegrationTest mostly focuses on the controller's modifications to zookeeper (and to some extent, the cluster) in reaction to the protocols. It especially focuses on controller epoch, leadership, and leader epoch updates from the controller, which the other integration tests don't pay attention to. ControllerIntegrationTest intentionally tries to directly trigger the protocols with zookeeper writes and ControlledShutdownRequest RPC instead of using the TopicCommand, ReassignPartitionsCommand, and PreferredReplicaLeaderElectionCommand.